### PR TITLE
Promise support in Events (addition to #9491)

### PR DESCRIPTION
**Pull request by johnspackman**
This PR has further commits to add to https://github.com/qooxdoo/qooxdoo/pull/9491, fixing some regressions introduced in that commit and adding unit tests

### CODE REVIEW

File: framework/source/class/qx/event/Utils.js [hunk 1]
```diff
@@ -103,26 +103,26 @@ qx.Class.define("qx.event.Utils", {
...
         }
         if (tracker.promise) {
           var self = this;
-          this.__push(tracker, tracker.promise.then(function() {
+          this.__push(tracker, tracker.promise.then(function(result) {
             if (tracker.rejected) {
               return null;
             }
-            var result = fn();
+            var result = fn(result);
```
Comment (derrell) [inline framework/source/class/qx/event/Utils.js]: This line is declaring `result`, so is then passing `undefined` (`result`) to `fn`. Something fishy going on here...

Comment (derrell) [reply framework/source/class/qx/event/Utils.js]: Maybe it's supposed to pass `tracker.result` ?

**Review by derrell — CHANGES_REQUESTED** [review]
In addition to my inline comments, it would be nice if the whitespace-changes-only files could be separated out into a separate PR.

File: framework/source/class/qx/event/Utils.js (hunk 1 above)
Comment (johnspackman) [reply framework/source/class/qx/event/Utils.js]: `result` is passed in so the bug is that it shouldn't be declaring it again with `var`; fixed

**Comment by johnspackman** [comment]
@derrell sorry about the white space changes, I wish git could ignore them - it's my IDE deciding it knows best :(

**Review by derrell — COMMENTED** [review]
The problem with the whitespace changes is that I *think* there are two files there with no changes other than whitespace, but some real change could just be obscured by those. I understand the IDE doing clean-up, and in general, clean-up is a good thing. I just want to know whether there actually are relevant changes in those two files. (Element.js and AbstractBubbling.js)

**Comment by johnspackman** [comment]
@derrell try adding `?w=1` to the URL!  It ignores whitespace in the diffs, more info here: https://github.com/blog/967-github-secrets

**Review by derrell — DISMISSED** [review]
?w=1 is the ticket! That's much more betterer!!! :-) Now I can see the substantive changes.

**Review by cboulanger — DISMISSED** [review]

**Review by level420 — DISMISSED** [review]

**Comment by level420** [comment]
@johnspackman I think it would be nice to have a more meaningfull title and some short description for this PR. currently the title is "Event promises 4" with no description. it would be nice if you could change this. thank you.

**Comment by level420** [comment]
@johnspackman could you please have a look into the failing travis job? It fails in two promise related tests:
```
[31mFirefox 58.0.0 (Linux 0.0.0) qx.test Promise:testAsyncEventHandlers FAILED[39m
	Asynchronous Test Error: Timeout reached before resume() was called.
		qx.core.AssertionError.constructor:326
		qx.dev.unit.TestResult:6727
		qx.dev.unit.TestResult.run:5892
		qx.dev.unit.TestResult:7404
		qx.event.Timer:1526
		qx.event.dispatch.Direct:3519
		qx.event.Utils.then:3304
		qx.event.dispatch.Direct:3396
		qx.event.dispatch.Direct.dispatchEvent:2195
		qx.Interface.wrappedFunction:13804
		qx.event.Manager.dispatchEvent:28476
		qx.event.Registration:8176
		qx.event.Utils:3304
		qx.event.Registration:8059
		qx.core.MEvent:3595
		qx.event.Timer:4045
		qx.event.GlobalError:3886
		qx.event.Timer:659
[31mFirefox 58.0.0 (Linux 0.0.0) qx.test Promise:testWaterfallBinding FAILED[39m
	Asynchronous Test Error: Timeout reached before resume() was called.
		qx.core.AssertionError.constructor:326
		qx.dev.unit.TestResult:6727
		qx.dev.unit.TestResult.run:5892
		qx.dev.unit.TestResult:7404
		qx.event.Timer:1526
		qx.event.dispatch.Direct:3519
		qx.event.Utils.then:3304
		qx.event.dispatch.Direct:3396
		qx.event.dispatch.Direct.dispatchEvent:2195
		qx.Interface.wrappedFunction:13804
		qx.event.Manager.dispatchEvent:28476
		qx.event.Registration:8176
		qx.event.Utils:3304
		qx.event.Registration:8059
		qx.core.MEvent:3595
		qx.event.Timer:4045
		qx.event.GlobalError:3886
		qx.event.Timer:659
```

**Comment by johnspackman** [comment]
No problem (I’m not sure I’ll be able to do it today though)

**Review by level420 — APPROVED** [review]

**Review by hkollmann — APPROVED** [review]

**Review by cboulanger — APPROVED** [review]

**Review by cajus — APPROVED** [review]

**Review by derrell — APPROVED** [review]

**Review by oetiker — APPROVED** [review]

**Comment by level420** [comment]
6 approvals, lots of business days, merging
thank you @johnspackman

**Pull request merged**